echopype/clean/impulse_noise.py

```diff
@@ -0,0 +1,463 @@
+    iax_ = np.arange(iax[0], iax[-1], m)
+    Sv_ = mask_transformation.oned(Sv, iax, iax_, 0, log_var=True)[0]
+
+    # resample back to full resolution
+    jax = np.arange(len(Sv[0]))
+    Sv_, mask_ = mask_transformation.full(Sv_, iax_, jax, iax, jax)
+
+    # perform side comparisons and combine masks in one unique mask
+    mask = np.zeros_like(Sv, dtype=bool)
+    for i in n:
+        dummy = np.zeros((iax.shape[0], i))
+        dummy[:] = np.nan
+        forward = Sv_ - np.c_[Sv_[:, i:], dummy]
+        backward = Sv_ - np.c_[dummy, Sv_[:, 0:-i]]
+        maskf = np.ma.masked_greater(forward, thr).mask
+        maskb = np.ma.masked_greater(backward, thr).mask
+        mask = mask | (maskf & maskb)
+
+    # get second mask indicating valid samples in IN mask
+    mask_[:, 0 : max(n)] = True
+    mask_[:, -max(n) :] = True
+
+    return mask, mask_
+
+
+def _wang(
+    Sv: np.ndarray,
+    thr: Tuple[float, float],
+    erode: List[Tuple[int, int]],
+    dilate: List[Tuple[int, int]],
+    median: List[Tuple[int, int]],
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Clean impulse noise from Sv data following the method described by:
+
+        Wang et al. (2015) ’A noise removal algorithm for acoustic data with
+        strong interference based on post-processing techniques’, CCAMLR
+        SG-ASAM: 15/02.
+
+    This algorithm runs different cycles of erosion, dilation, and median
+    filtering to clean impulse noise from Sv. Note that this function
+    returns a clean/corrected Sv array, instead of a boolean array indicating
+    the occurrence of impulse noise.
+
+    Parameters
+    ----------
+        Sv     (float)    : 2D numpy array with Sv data (dB).
+        thr    (int/float): 2-element tuple with bottom/top Sv thresholds (dB)
+        erode  (int)      : list of 2-element tuples indicating the window's
+                            size for each erosion cycle.
+        dilate (int)      : list of 2-element tuples indicating the window's
+                            size for each dilation cycle.
+        median (int)      : list of 2-element tuples indicating the window's
+                            size for each median filter cycle.
+
+    Returns
+    -------
+        float             : 2D array with clean Sv data.
+        bool              : 2D array with mask indicating valid clean Sv data.
+    """
+
+    # set weak noise and strong interference as vacant samples (-999)
+    Sv_thresholded = Sv.copy()
+    Sv_thresholded[(Sv < thr[0]) | (Sv > thr[1])] = -999
+
+    # remaining weak interferences will take neighbouring vacant values
+    # by running erosion cycles
+    Sv_eroded = Sv.copy()
+    for e in erode:
+        Sv_eroded = erosion(Sv_thresholded, np.ones(e))
+
+    # the last step might have turned interferences inside biology into vacant
+    # samples, this is solved by running dilation cycles
+    Sv_dilated = Sv_eroded.copy()
+    for d in dilate:
+        Sv_dilated = dilation(Sv_dilated, np.ones(d))
```

Comment:
[dask_image.ndmorph package](https://image.dask.org/en/latest/dask_image.ndmorph.html) seems to have wrapped `scipy.ndimage._morphology.binary_erosion`. Could you test if it takes delayed dask arrays correctly (so that it can deal with large, lazy-loaded datasets)?

Reply:
Unfortunately, we use grayscale erosion rather than binary erosion (and after verifying with the original paper, we need it) so this can't be ported in dask and will have to stay as numpy operations.